allennlp/semparse/contexts/atis_tables.py

```diff
@@ -0,0 +1,437 @@
+
+The first element is the starting symbol. We initialize ``col_ref``, ``table_ref``,
+``table_name``, ``number``, ``string`` to empty productions first, we will fill in
+``col_ref``, ``table_ref``, ``table_name`` based on the dataset and ``number`` and
+``string`` based on the utterances
+"""
+
+SQL_GRAMMAR_STR = r"""
+    stmt                = query ws ";" ws
+    query               = (ws "(" ws "SELECT" ws distinct ws select_results ws "FROM" ws table_refs ws where_clause ws ")" ws) /
+                          (ws "SELECT" ws distinct ws select_results ws "FROM" ws table_refs ws where_clause ws)
+                        
+    select_results      = col_refs / agg
+
+    agg                 = agg_func ws "(" ws col_ref ws ")"
+    agg_func            = "MIN" / "min" / "MAX" / "max" / "COUNT" / "count"
+
+    col_refs            = (col_ref ws "," ws col_refs) / (col_ref)
+    table_refs          = (table_name ws "," ws table_refs) / (table_name)
+
+    where_clause        = ("WHERE" ws "(" ws conditions ws ")" ws) / ("WHERE" ws conditions ws)
+    
+    conditions          = (condition ws conj ws conditions) / 
+                          (condition ws conj ws "(" ws conditions ws ")") /
+                          ("(" ws conditions ws ")" ws conj ws conditions) /
+                          ("(" ws conditions ws ")") /
+                          (not ws conditions ws ) /
+                          condition
+    condition           = in_clause / ternaryexpr / biexpr
+
+    in_clause           = (ws col_ref ws "IN" ws query ws)
+
+    biexpr              = ( col_ref ws binaryop ws value) / (value ws binaryop ws value) / ( col_ref ws "LIKE" ws string)
+    binaryop            = "+" / "-" / "*" / "/" / "=" /
+                          ">=" / "<=" / ">" / "<"  / "is" / "IS"
+
+    ternaryexpr         = (col_ref ws not "BETWEEN" ws value ws and value ws) /
+                          (col_ref ws "BETWEEN" ws value ws and value ws)
+
+    value               = (not ws pos_value) / (pos_value)
+    pos_value           = ("ALL" ws query) / ("ANY" ws query) / number / boolean / col_ref / string / agg_results / "NULL"
+
+    agg_results         = (ws "("  ws "SELECT" ws distinct ws agg ws "FROM" ws table_name ws where_clause ws ")" ws) /
+                          (ws "SELECT" ws distinct ws agg ws "FROM" ws table_name ws where_clause ws)
+
+    boolean             = "true" / "false"
+
+    ws                  = ~"\s*"i
+
+    conj                = and / or
+    and                 = "AND" ws
+    or                  = "OR" ws
+    not                 = ("NOT" ws ) / ("not" ws)
+    asterisk            = "*"
+    distinct            = ("DISTINCT") / ("")
+
+    col_ref             =  ""
+    table_ref           =  ""
+    table_name          =  ""
+    number              =  ""
+    string              =  ""
+
+"""
+
+class ConversationContext():
+    """
+    A ``ConversationContext`` represents the interaction in which an utterance occurs.
```

Comment:
Yes, I agree, the generic SQL functionality should be separated out. I was planning to leave that for a later PR since I haven't tested this on the other datasets yet. I'll separate it out now so it is easier/cleaner in the future to reuse this for other SQL datasets.